Modules/_datetimemodule.c

```diff
@@ -1939,6 +1949,22 @@ wrap_strftime(PyObject *object, PyObject *format, PyObject *timetuple,
             ptoappend = PyBytes_AS_STRING(freplacement);
             ntoappend = PyBytes_GET_SIZE(freplacement);
         }
+#ifdef NORMALIZE_CENTURY
+        else if (ch == 'Y' || ch == 'G') {
+            /* 0-pad year with century as necessary */
+            if (ch == 'G') {
+                format = PyUnicode_FromString("%G");
+                result = PyObject_CallFunctionObjArgs(strftime, format,
+                                                      timetuple, NULL);
+                year = atoi(PyUnicode_AsUTF8(result));
```

Comment:
`result` can be NULL in case of error. There is also no guarantee that it is a string, and that it is UTF-8 encodable, and that `PyUnicode_AsUTF8` will not fail due to memory error. So in this code you need to check result for NULL, check that it is a string and raise an exception if it is not, check the returned value of `PyUnicode_AsUTF8()` for NULL. BTW, `atoi()` can fail, and there is no portable way to check for its error. You should use `strtol()` instead.

You can also use other ways. For example, using `PyObject_CallFunction(strftime, "sO", "%G", timetuple)` saves few lines for creating, checking and destroying `format`. But it is slower, so you should measure and decide. You can use `PyNumber_Long()` + `PyLong_AsLong()` to convert from the Python string to C long. It may need less checks, but it is slower. Again, you should measure and decide whether the cost of saving several lines of code is worth the slowdown.

Reply:
Many thanks for your careful review and guidance. `PyObject_CallFunction` and `PyNumber_Long` + `PyLong_AsLong()` do make the code a lot cleaner. I've gone back and forth timing my original attempt and the new code with either of the two replacement functions, and could not produce any meaningful performance difference (as in any difference appeared to be a result of system load flunctuations at the different times of running), so I've gone with your suggestions. Thanks again.